### PR TITLE
[Feature] enable async transfer in NodeDataLoader for homograph

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Description

part of https://github.com/dmlc/dgl/issues/3294.
In this part, async data transfer is enabled in NodeDataLoader for homograph.
feature/label prefetch is also enabled.
async_load is independent of load_input/output.
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/dataloading/pytorch/dataloader.py [hunk 1]
```diff
@@ -422,6 +491,17 @@ class NodeDataLoader:
...
 
         Only effective when :attr:`use_ddp` is True.
+    load_input : dict[tag, Tensor], optional
+        The tensors will be sliced according to ``blocks[0].srcdata[dgl.NID]``
+        and will be attached to ``blocks[0].srcdata``.
+    load_output : dict[tag, Tensor], optional
+        The tensors will be sliced according to ``blocks[-1].dstdata[dgl.NID]``
+        and will be attached to ``blocks[-1].dstdata``.
+    async_load : boolean, optional
+        If True, data including graph, sliced tensors will be transferred
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply python/dgl/dataloading/pytorch/dataloader.py:500]: I find this is caused by frontier is already transferred to gpu in `sampler.sample_block()` on default stream. I disable it and graph transfer happens on the stream we created.

File: python/dgl/dataloading/pytorch/dataloader.py [hunk 2]
```diff
@@ -307,23 +310,89 @@ def _to_device(data, device):
...
+
+def _background_node_dataloader(dl_iter, g, device, results, load_input, load_output):
+    try:
+        while True:
+            dev = None
+            if device.type == 'cuda':
+                dev = device
+            elif g.device.type == 'cuda':
+                dev = g.device
+            stream = th.cuda.Stream(device=dev)
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply python/dgl/dataloading/pytorch/dataloader.py]: yes.

File: python/dgl/dataloading/pytorch/dataloader.py [hunk 3]
```diff
@@ -307,23 +310,89 @@ def _to_device(data, device):
...
-        _restore_blocks_storage(result_[-1], self.node_dataloader.collator.g)
-
-        result = [_to_device(data, self.device) for data in result_]
-        return result
+        res = ()
+        if self.async_load:
+            (res, _), stream = self.results.get()
+            if res is None:
+                raise StopIteration
+            stream.synchronize()
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply python/dgl/dataloading/pytorch/dataloader.py]: you're right.

**Review by nv-dlasalle — APPROVED** [review]
Couple of minor comments, overall looks good to me.

File: python/dgl/dataloading/pytorch/dataloader.py [hunk 4]
```diff
@@ -307,23 +310,88 @@ def _to_device(data, device):
...
+    except StopIteration:
+        results.put((None, None, None))
+
+
 class _NodeDataLoaderIter:
     def __init__(self, node_dataloader):
         self.device = node_dataloader.device
         self.node_dataloader = node_dataloader
         self.iter_ = iter(node_dataloader.dataloader)
+        self.async_load = node_dataloader.async_load and th.cuda.is_available()
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply python/dgl/dataloading/pytorch/dataloader.py:365]: I tried to enable async transfer between cpu to gpu and vice versa. But actually no profiling is done for gpu2cpu. So let's check as you said for now.

File: python/dgl/dataloading/pytorch/dataloader.py [hunk 5]
```diff
@@ -422,6 +490,17 @@ class NodeDataLoader:
         :class:`torch.utils.data.distributed.DistributedSampler`.
 
         Only effective when :attr:`use_ddp` is True.
+    load_input : dict[tag, Tensor], optional
+        The tensors will be sliced according to ``blocks[0].srcdata[dgl.NID]``
+        and will be attached to ``blocks[0].srcdata``.
+    load_output : dict[tag, Tensor], optional
+        The tensors will be sliced according to ``blocks[-1].dstdata[dgl.NID]``
+        and will be attached to ``blocks[-1].dstdata``.
+    async_load : boolean, optional
```
[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply python/dgl/dataloading/pytorch/dataloader.py:499]: yes. I added this note into the description part of `async_load` for user's awareness.
